File: aes-ni/oxcaml/benchmark_aesni.ml
(** AES-NI benchmark using OCaml C bindings.
    Stubs follow the Cryptokit pattern by Xavier Leroy. *)

open Aesni

let benchmark_file filename =
  if check_available () = 0 then begin
    Printf.eprintf "AES-NI not supported on this CPU\n";
    exit 1
  end;

  let ic = open_in_bin filename in
  let original_len = in_channel_length ic in
  let data = Bytes.create original_len in
  really_input ic data 0 original_len;
  close_in ic;

  let original_data = Bytes.copy data in

  let padded_len = ((original_len + 15) / 16) * 16 in
  let padded = Bytes.make padded_len '\000' in
  Bytes.blit data 0 padded 0 original_len;

  (* Fixed 16-byte key matching the C benchmark *)
  let key = Bytes.of_string
    "\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0a\x0b\x0c\x0d\x0e\x0f" in

  let ckey_enc = cook_encrypt_key key in
  let ckey_dec = cook_decrypt_key key in

  let nblocks   = padded_len / 16 in
  let encrypted = Bytes.create padded_len in
  let decrypted = Bytes.create padded_len in

  (* Encrypt: OCaml loop, one C call per 16-byte block (cryptokit pattern) *)
  let enc_start = Unix.gettimeofday () in
  for i = 0 to nblocks - 1 do
    let off = i * 16 in
    encrypt_block ckey_enc padded off encrypted off
  done;
  let enc_end = Unix.gettimeofday () in

  (* Decrypt *)
  let dec_start = Unix.gettimeofday () in
  for i = 0 to nblocks - 1 do
    let off = i * 16 in
    decrypt_block ckey_dec encrypted off decrypted off
  done;
  let dec_end = Unix.gettimeofday () in

  let enc_time = enc_end -. enc_start in
  let dec_time = dec_end -. dec_start in
  let mb       = float_of_int original_len /. (1024.0 *. 1024.0) in
  let size_mb  = int_of_float mb in

  (* Verify round-trip *)
  let ok = ref true in
  for i = 0 to original_len - 1 do
    if Bytes.get original_data i <> Bytes.get decrypted i then ok := false
  done;
  if not !ok then begin
    Printf.eprintf "Verification: FAILED\n";
    exit 1
  end;

  Printf.printf "%d,%.6f,%.6f,%.2f,%.2f\n"
    size_mb enc_time dec_time (mb /. enc_time) (mb /. dec_time)

let () =
  if Array.length Sys.argv <> 2 then begin
    Printf.eprintf "Usage: %s <input_file>\n" Sys.argv.(0);
    exit 1
  end;
  benchmark_file Sys.argv.(1)

File: aes-ni/c/benchmark_aesni.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define EXPORT extern
#include "aesni.h"

static double get_time(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec + ts.tv_nsec / 1e9;
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <input_file>\n", argv[0]);
        return 1;
    }

    if (!aesni_check_available()) {
        fprintf(stderr, "AES-NI not supported on this CPU\n");
        return 1;
    }

    FILE *fp = fopen(argv[1], "rb");
    if (!fp) {
        perror("fopen");
        return 1;
    }

    fseek(fp, 0, SEEK_END);
    long msg_len = ftell(fp);
    rewind(fp);

    long padded_len = ((msg_len + 15) / 16) * 16;

    unsigned char *message = calloc(padded_len, 1);
    if (!message) {
        fprintf(stderr, "Memory allocation failed\n");
        fclose(fp);
        return 1;
    }
    fread(message, 1, msg_len, fp);
    fclose(fp);

    /* Fixed 16-byte key for AES-128 */
    static const unsigned char key[16] = {
        0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
        0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f
    };

    /* Key schedule buffers: 15 round keys × 16 bytes covers AES-128/192/256 */
    unsigned char ckey_enc[15 * 16];
    unsigned char ckey_dec[15 * 16];

    int nr_enc = aesniKeySetupEnc(ckey_enc, key, 128);
    int nr_dec = aesniKeySetupDec(ckey_dec, key, 128);

    unsigned char *encrypted = malloc(padded_len);
    unsigned char *decrypted = malloc(padded_len);
    if (!encrypted || !decrypted) {
        fprintf(stderr, "Memory allocation failed\n");
        free(message);
        return 1;
    }

    double enc_start = get_time();
    for (long i = 0; i < padded_len; i += 16)
        aesniEncrypt(ckey_enc, nr_enc, message + i, encrypted + i);
    double enc_end = get_time();

    double dec_start = get_time();
    for (long i = 0; i < padded_len; i += 16)
        aesniDecrypt(ckey_dec, nr_dec, encrypted + i, decrypted + i);
    double dec_end = get_time();

    int ok = (memcmp(message, decrypted, msg_len) == 0);

    double enc_time = enc_end - enc_start;
    double dec_time = dec_end - dec_start;
    double size_mb  = (double)msg_len / (1024.0 * 1024.0);

    printf("Message length      : %ld bytes\n",  msg_len);
    printf("Encryption time     : %.6f sec\n",   enc_time);
    printf("Decryption time     : %.6f sec\n",   dec_time);
    printf("Encryption speed    : %.2f MB/s\n",  size_mb / enc_time);
    printf("Decryption speed    : %.2f MB/s\n",  size_mb / dec_time);
    printf("Verification        : %s\n",          ok ? "PASSED" : "FAILED");

    free(message);
    free(encrypted);
    free(decrypted);
    return 0;
}
